README.md
# swarmerode

[![Build Status](https://img.shields.io/travis/Runnable/swarmerode/master.svg?style=flat-square)](https://travis-ci.org/Runnable/swarmerode)
[![Test Coverage](https://img.shields.io/coveralls/Runnable/swarmerode.svg?style=flat-square)](https://coveralls.io/github/Runnable/swarmerode)
[![Dependency Status](https://img.shields.io/david/Runnable/swarmerode.svg?style=flat-square)](https://david-dm.org/Runnable/swarmerode)
[![devDependency Status](https://img.shields.io/david/dev/Runnable/swarmerode.svg?style=flat-square)](https://david-dm.org/Runnable/swarmerode#info=devDependencies)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com/)

Swarmerode is a tool to extend [`dockerode`](https://www.npmjs.com/package/dockerode) with nice swarm helper functions.

## Example

```javascript
var Swarmerode = require('swarmerode')
var Dockerode = require('dockerode')

// extend Dockerode w/ Swarm functions.
Dockerode = Swarmerode(Dockerode)

var swarmClient = new Dockerode(/* opts */)
swarmClient.swarmHostExists('10.0.0.1:2375', function (err, hostExists) {
  if (err) { throw err }
  console.log(
    hostExists
      ? 'The given host exists!'
      : 'The given host was not found.'
  )
})
```


CHANGELOG.md
## Change Log

### v1.0.0 (2016/04/18 21:53 +00:00)
- [#5](https://github.com/Runnable/swarmerode/pull/5) Changed to support swarm 1.2.0

### v0.2.0 (2016/03/10 17:12 +00:00)
- [#2](https://github.com/Runnable/swarmerode/pull/2) add github-changes (@bkendall)

### v0.1.0 (2016/01/21 21:40 +00:00)
- [#1](https://github.com/Runnable/swarmerode/pull/1) initial implementation and tests (@bkendall)


test/index.js
'use strict'

var assert = require('chai').assert
var clone = require('101/clone')
var sinon = require('sinon')

var exampleHosts = [ '10.0.0.1:4242', '10.0.0.2:4242', '10.0.0.3:4242' ]
var swarmInfoMock = require('./fixtures/swarm-info')
var testInfo = swarmInfoMock([{ host: exampleHosts[0] }, { host: exampleHosts[1] }, { host: exampleHosts[2] }])
var Swarmerode = require('../')

describe('Swarmerode', function () {
  var MockClass
  var instance
  beforeEach(function () {
    MockClass = function () {}
    MockClass.prototype.info = function (cb) { cb(null, clone(testInfo)) }
    MockClass = Swarmerode(MockClass)
    instance = new MockClass()
  })

  it('should extend a given class', function () {
    assert.isFunction(MockClass.prototype.swarmHosts)
    assert.isFunction(MockClass.prototype.swarmHostExists)
    assert.isFunction(instance.swarmHosts)
    assert.isFunction(instance.swarmHostExists)
  })

  it('should not allow being instantiate directly', function () {
    var SwarmerodeClass = Swarmerode._Swarmerode
    assert.throws(
      function () { return new SwarmerodeClass() },
      Error,
      /cannot instantiate swarmerode directly/i
    )
  })

  it('should not override existing prototype properties', function () {
    var sampleObject = { foo: 'bar' }
    var SampleClass = function () {}
    SampleClass.prototype.swarmHostExists = sampleObject
    SampleClass = Swarmerode(SampleClass)
    var sampleInstance = new SampleClass()
    assert.isObject(sampleInstance.swarmHostExists)
    assert.equal(sampleInstance.swarmHostExists, sampleObject)
  })

  it('should not override existing prototype functions', function () {
    var sampleFunction = function () {}
    var SampleClass = function () {}
    SampleClass.prototype.swarmHostExists = sampleFunction
    SampleClass = Swarmerode(SampleClass)
    var sampleInstance = new SampleClass()
    assert.isFunction(sampleInstance.swarmHostExists)
    assert.equal(sampleInstance.swarmHostExists, sampleFunction)
  })

  it('should not override existing class properties', function () {
    var sampleObject = { foo: 'bar' }
    var SampleClass = function () {}
    SampleClass.swarmHostExists = sampleObject
    SampleClass = Swarmerode(SampleClass)
    assert.isObject(SampleClass.swarmHostExists)
    assert.equal(SampleClass.swarmHostExists, sampleObject)
  })

  it('should not override existing class functions', function () {
    var sampleFunction = function () {}
    var SampleClass = function () {}
    SampleClass.swarmHostExists = sampleFunction
    SampleClass = Swarmerode(SampleClass)
    assert.isFunction(SampleClass.swarmHostExists)
    assert.equal(SampleClass.swarmHostExists, sampleFunction)
  })

  describe('swarmHosts', function () {
    it('should return any error from docker', function (done) {
      var error = new Error('robot')
      sinon.stub(MockClass.prototype, 'info').yieldsAsync(error)
      instance.swarmHosts(function (err) {
        assert.equal(err, error)
        done()
      })
    })

    it('should return a list of swarm hosts', function (done) {
      instance.swarmHosts(function (err, hosts) {
        assert.isNull(err)
        assert.lengthOf(hosts, 3)
        assert.includeMembers(hosts, exampleHosts)
        done()
      })
    })
  })

  describe('swarmHostExists', function () {
    it('should return any error from docker', function (done) {
      var error = new Error('robot')
      sinon.stub(MockClass.prototype, 'info').yieldsAsync(error)
      instance.swarmHostExists('10.0.0.1:4242', function (err) {
        assert.equal(err, error)
        done()
      })
    })

    it('should return a boolean if host in swarm (true)', function (done) {
      instance.swarmHostExists('10.0.0.1:4242', function (err, exists) {
        assert.isNull(err)
        assert.isTrue(exists)
        done()
      })
    })

    it('should return a boolean if host in swarm (false)', function (done) {
      instance.swarmHostExists('10.0.0.4:4242', function (err, exists) {
        assert.isNull(err)
        assert.isFalse(exists)
        done()
      })
    })
  })

  describe('_parseSwarmSystemStatus', function () {
    it('should format SystemStatus correctly', function (done) {
      var coolNode = {
        Labels: 'env=test, hd=ssd',
        Containers: 100,
        nodeName: '  cool.node',
        host: '10.42.42.42:4242'
      }
      var uncoolNode = {
        Labels: 'env=prod, hd=disk',
        Containers: 4,
        nodeName: '  un.cool.node',
        host: '10.7.7.7:4242'
      }
      var testHosts = swarmInfoMock([coolNode, uncoolNode])

      var out = Swarmerode._Swarmerode._parseSwarmSystemStatus(testHosts.SystemStatus)
      assert.equal(out.Role, 'primary')
      assert.equal(out.Strategy, 'spread')
      assert.equal(out.Filters, 'health, port, dependency, affinity, constraint')
      assert.isNumber(out.Nodes)
      assert.equal(out.Nodes, 2)

      assert.equal(out.ParsedNodes['cool.node'].Host, coolNode.host)
      assert.isNumber(out.ParsedNodes['cool.node'].Containers)
      assert.equal(out.ParsedNodes['cool.node'].Containers, 100)
      assert.equal(out.ParsedNodes['cool.node'].Status, 'Healthy')
      assert.equal(out.ParsedNodes['cool.node'].ReservedCpus, '0 / 1')
      assert.equal(out.ParsedNodes['cool.node'].ReservedMem, '10 GiB / 1.021 GiB')
      assert.equal(out.ParsedNodes['cool.node'].Error, '(none)')
      assert.equal(out.ParsedNodes['cool.node'].UpdatedAt, '2016-03-08T19:02:41Z')
      assert.equal(out.ParsedNodes['cool.node'].Labels.env, 'test')
      assert.equal(out.ParsedNodes['cool.node'].Labels.hd, 'ssd')

      assert.equal(out.ParsedNodes['un.cool.node'].Host, uncoolNode.host)
      assert.isNumber(out.ParsedNodes['un.cool.node'].Containers)
      assert.equal(out.ParsedNodes['un.cool.node'].Status, 'Healthy')
      assert.equal(out.ParsedNodes['un.cool.node'].Containers, 4)
      assert.equal(out.ParsedNodes['un.cool.node'].ReservedCpus, '0 / 1')
      assert.equal(out.ParsedNodes['un.cool.node'].ReservedMem, '10 GiB / 1.021 GiB')
      assert.equal(out.ParsedNodes['un.cool.node'].Error, '(none)')
      assert.equal(out.ParsedNodes['un.cool.node'].UpdatedAt, '2016-03-08T19:02:41Z')
      assert.equal(out.ParsedNodes['un.cool.node'].Labels.env, 'prod')
      assert.equal(out.ParsedNodes['un.cool.node'].Labels.hd, 'disk')
      done()
    })
  }) // end _parseSwarmSystemStatus
})


index.js
'use strict'

var clone = require('101/clone')
var debug = require('debug')('swarmerode')
var exists = require('101/exists')

/*
 * Swarmerode Class constructor. Not really to be used but as a placeholder for
 * the functions with which we want to extend dockerode.
 */
function Swarmerode () {
  throw new Error('Cannot instantiate Swarmerode directly')
}

/**
 * Return a list of swarm hosts (e.g. `[ '10.0.0.1:2375', '10.0.0.2:2375' ]`).
 * @param {Function} cb Callback with signature (err, hosts).
 */
Swarmerode.prototype.swarmHosts = function (cb) {
  this.swarmInfo(function (err, data) {
    if (err) { return cb(err) }
    var nodes = data.parsedSystemStatus.ParsedNodes
    var hosts = Object.keys(nodes)
      .map(function (key) {
        return nodes[key].Host
      })
    cb(null, hosts)
  })
}

/**
 * Return a formated list of swarm nodes
 * @param {Function} cb Callback with signature (err, nodes).
 */
Swarmerode.prototype.swarmInfo = function (cb) {
  this.info(function (err, info) {
    if (err) { return cb(err) }
    info.parsedSystemStatus = Swarmerode._parseSwarmSystemStatus(info.SystemStatus)
    debug('swarm info %j', info)
    cb(null, info)
  })
}

/**
 * Return a boolean indicating if a swarm host exists.
 * @param {String} host Host address (e.g. `10.0.0.1:2375`).
 * @param {Function} cb Callback with signature (err, hostExists).
 */
Swarmerode.prototype.swarmHostExists = function (host, cb) {
  this.swarmHosts(function (err, hosts) {
    if (err) { return cb(err) }
    var index = hosts.indexOf(host)
    cb(null, index !== -1)
  })
}

/**
 * Parse and transform raw swarm data into the proper JSON.
 * format: {
 *   Role: 'primary'
 *   Strategy: 'spread'
 *   Filters: 'health, port, dependency, affinity, constraint'
 *   Nodes: 42
 *   ParsedNodes: {
 *     <nodeHostname>: {
 *       Host: '10.0.0.1:4242,
 *       Containers: 5,
 *       ReservedCpus: '0 / 1',
 *       ReservedMem: '10 GiB / 1.021 GiB',
 *       Labels: {
 *         env: 'prod',
 *         provider: 'virtualbox'
 *       },
 *       Error: '(none)',
 *       UpdatedAt: '2016-03-08T19:02:41Z'
 *     }
 *   }
 * }
 * @param {Array} - array data response from swarm info
 * @return array of json objects with swarm data for each node
 */
Swarmerode._parseSwarmSystemStatus = function (systemStatus) {
  systemStatus = clone(systemStatus)
  // the first 4 fields are Role, Strategy, Filters, Nodes respectively
  var formatted = {
    Role: systemStatus.shift()[1],
    Strategy: systemStatus.shift()[1],
    Filters: systemStatus.shift()[1],
    Nodes: parseInt(systemStatus.shift()[1], 10),
    ParsedNodes: {}
  }

  for (var i = 0; i < formatted.Nodes; i++) {
    formatted.ParsedNodes[systemStatus[0][0].trim()] = {
      Host: systemStatus.shift()[1],
      Status: systemStatus.shift()[1],
      Containers: parseInt(systemStatus.shift()[1], 10),
      ReservedCpus: systemStatus.shift()[1],
      ReservedMem: systemStatus.shift()[1],
      Labels: parseLabels(systemStatus.shift()[1]),
      Error: systemStatus.shift()[1],
      UpdatedAt: systemStatus.shift()[1],
      ServerVersion: systemStatus.shift()[1]
    }
  }

  function parseLabels (labelString) {
    var labelsTokens = labelString.split(',')
    var labels = {}
    labelsTokens.forEach(function (labelToken) {
      var pair = labelToken.split('=').map(function (s) {
        return s.trim()
      })
      labels[pair[0]] = pair[1]
    })
    return labels
  }

  return formatted
}

/**
 * Extend a given Class's prototype with the swarm functions.
 * @param {Function} Class A 'Class' of which to extend the prototype
 * @returns {Function} The given Class with swarm functions.
 */
module.exports = function (Class) {
  Object.getOwnPropertyNames(Swarmerode.prototype).forEach(function (key) {
    if (key === 'constructor') { return }
    if (exists(Class.prototype[key])) { return }
    Class.prototype[key] = Swarmerode.prototype[key]
  })
  return Class
}

// exposing the class for testing purposes only
module.exports._Swarmerode = Swarmerode
